alf/environments/alf_wrappers.py

```diff
@@ -465,3 +481,285 @@ def reward_spec(self):
+            "environment should have exactly two keys: 'task_id' and 'action'")
+        self._action_spec = env.action_spec()['action']
+        self._num_tasks = num_tasks
+        self._task_names = task_names
+        self._env_info_spec = copy.copy(env.env_info_spec())
+        self._env_info_spec.update(
+            self._add_task_names({
+                'curriculum_task_count': [alf.TensorSpec(())] * num_tasks,
+                'curriculum_task_score': [alf.TensorSpec(())] * num_tasks,
+                'curriculum_task_prob': [alf.TensorSpec(())] * num_tasks
+            }))
+        self._zero_curriculum_info = self._add_task_names({
+            'curriculum_task_count': [torch.zeros(batch_size, device='cpu')] *
+                                     num_tasks,
+            'curriculum_task_score': [torch.zeros(batch_size, device='cpu')] *
+                                     num_tasks,
+            'curriculum_task_prob': [torch.zeros(batch_size, device='cpu')] *
+                                    num_tasks
+        })
+        self._progress_favor = progress_favor
+        self._current_score_update_rate = current_score_update_rate
+        self._past_score_update_rate = past_score_update_rate
+        self._warmup_period = warmup_period * num_tasks
+        self._scale = math.log(progress_favor)
+        self._total_count = 0
+        self._current_scores = torch.zeros(num_tasks, device='cpu')
+        self._past_scores = torch.zeros(num_tasks, device='cpu')
+        self._task_probs = torch.ones(num_tasks, device='cpu') / num_tasks
+        self._task_counts = torch.zeros(num_tasks, device='cpu')
+
+        self._current_task_ids = self._sample_tasks(batch_size)
+
+    def _add_task_names(self, info):
+        for k, v in info.items():
+            info[k] = dict(zip(self._task_names, v))
+        return info
+
+    def _sample_tasks(self, num_samples):
+        return torch.multinomial(
+            self._task_probs, num_samples=num_samples, replacement=True)
+
+    def _update_curriculum(self, task_ids, task_scores):
+        for task_id, task_score in zip(task_ids, task_scores):
+            self._total_count += 1
+            self._task_counts[task_id] += 1
+            self._current_scores[
+                task_id] += self._current_score_update_rate * (
+                    task_score - self._current_scores[task_id])
+            self._past_scores[task_id] += self._past_score_update_rate * (
+                task_score - self._past_scores[task_id])
+
+        progresses = (self._current_scores - self._past_scores).relu()
+        max_progress = progresses.max()
+        progresses = progresses / (max_progress + 1e-30)
+        # Gradually increase scale from 0 to self._scale so that we tend to do
+        # random smapling of the environments initially
+        scale = self._scale * min(1, self._total_count / self._warmup_period)
+        self._task_probs = F.softmax(scale * progresses, dim=0)
```

Comment:
This task_probs calculation is kind of unintuitive. The progresses are basically ratios but treated as logits after scaling?

Reply:
progresses are normalized to [0, 1]